000X-B-C-I-R-S-error-handling-codes.md

```diff
@@ -0,0 +1,309 @@
+|403  |	PERMISSION_DENIED |	Client does not have sufficient permission. This can happen because the OAuth token does not have the right scopes, the client doesn't have permission, or the API has not been enabled. |
+|404  |	NOT_FOUND |	A specified resource is not found. |
+|409  |	ABORTED	| Concurrency conflict, such as read-modify-write conflict. |
+
+
+### ErrorInfo (Required)
+The properties of the **type.googleapis.com/google.rpc.ErrorInfo** will be populated as following:
+
+- **Reason** - A combination of prefix from prefix of the table below plus the error condition code.
+    
+    Example: "DAPR_STATE_" + "ETAG_MISMATCH"
+
+- **Domain** - With the value `dapr.io`.
+
+- **Metadata** - A key/value map/dictionary data relevant to the error condition.
+
+`**Note:**` The metadata property **isRetriable** with a truthable value("true", "false", "True", "False", "TRUE", "FALSE", "1", "0") is required. 
+
+### ResourceInfo (Optional)
+The properties of the **type.googleapis.com/google.rpc.ResourceInfo** will be populated as following:
+
+- **ResourceType** - The building block type.
+   
+   Example: "StateStore"
+
+- **ResourceName** - The component name.
+   
+   Example: "REDIS"
+
+- **Owner**   - The owner of the component.
+
+- **Description** - Resource descrpition or error details.
+
+
+### Error Details Prefixes
+The following tables shows the propsosed error codes prefixes used in the **reason** for the **google.rpc.ErrorInfo** for various Dapr building blocks:
+
+
+**INIT**
+| Dapr Module | Prefix |
+| ----------- | ----------- |
+| CLI         | DAPR_CLI_INIT_* |
+| Self-hosted | DAPR_SELF_HOSTED_INIT_* |
+| K8S         | DAPR_K8S_INIT_* |
+| Invoke      | DAPR_INVOKE_INIT_* |
+
+**RUNTIME**
+| Dapr Module | Prefix |
+| ----------- | ----------- |
+| CLI         | DAPR_CLI_RUNTIME_* |
+| Self-hosted | DAPR_SELF_HOSTED_RUNTIME_* |
+| dapr-2-dapr(HTTP) | DAPR_HTTP_RUNTIME_* |
+| dapr-2-dapr(gRPC) | DAPR_GRPC_RUNTIME_* |
+
+**COMPONENTS**
+| Dapr Module | Prefix |
+| ----------- | ----------- |
+| PubSub              | DAPR_PUBSUB_* |
+| StateStore          | DAPR_STATE_* |
+| Bindings            | DAPR_BINDING_* |
+| SecretStore         | DAPR_SECRET_* |
+| ConfigurationStore  | DAPR_CONFIGURATION_* |
+| Lock                | DAPR_LOCK_* |
+| NameResolution      | DAPR_NAME_RESOLUTION_* |
+| Middleware          | DAPR_MIDDLEWARE_*|
+
+
+The following snippet shows an error status returned due to a `ETAG_MISMATCH` error condition. The **reason** is populated with `PREFIX+ERROR_CONDITION`:
+
+```json
+{
+  "code": 3,
+  "message":  "possible etag mismatch. error from state store",
+  "details": [
+    {
+      "@type": "type.googleapis.com/google.rpc.ErrorInfo",
+      "reason": "DAPR_STATE_ETAG_MISMATCH",
+      "domain": "dapr.io",
+      "metadata": {
+        "key": "myapp||name"
+      }
+    },
+    {
+      "@type": "type.googleapis.com/google.rpc.ResourceInfo",
+      "resource_type": "StateStore",
```

Comment:
Some keys use camelCase, some use snake_case. We should be consistent

Also see my comment above. I think a single `resource_name` = `state.redis/v1` may be more aligned with how we refer to the comoponents everywhere (note that some components have a dot in the name too, such as `state.azure.cosmosdb`, and we should include the version too)

Reply:
The `keys`  for the JSON payload are auto-generated as part of the Marshaling implemented by the Protobuf stub of the [error_details.proto](https://github.com/googleapis/googleapis/blob/master/google/rpc/error_details.proto) 

if you look at the annotations, you'll see the `resource_type` as the JSON key to generate/read.
```
ResourceType string `protobuf:"bytes,1,opt,name=resource_type,json=resourceType,proto3" json:"resource_type,omitempty"`
```